### PR TITLE
boards/arm/stm32/nucleo-f446re: added support for PWM

### PR DESCRIPTION
**Pull request by michallenc**
## Summary
Added PWM support and example configuration for Nucleo F446RE board.

## Testing
Tested with pwm example application, output measured with osclloscope.

### CODE REVIEW

File: boards/arm/stm32/nucleo-f446re/configs/pwm/defconfig [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+CONFIG_NSH_READLINE=y
+CONFIG_PREALLOC_TIMERS=4
+CONFIG_PWM=y
+CONFIG_RAM_SIZE=131072
+CONFIG_RAM_START=0x20000000
+CONFIG_RAW_BINARY=y
+CONFIG_RR_INTERVAL=200
+CONFIG_SCHED_WAITPID=y
+CONFIG_SDCLONE_DISABLE=y
+CONFIG_SPI=y
```
Comment (saramonteiro) [inline boards/arm/stm32/nucleo-f446re/configs/pwm/defconfig:45]: we don't need SPI here
This an nit that has been propagating in several defconfs

File: boards/arm/stm32/nucleo-f446re/src/nucleo-f446re.h [hunk 2]
```diff
@@ -330,4 +330,16 @@ int board_ajoy_initialize(void);
 int stm32_foc_setup(void);
 #endif
 
+/****************************************************************************
+ * Name: stm32_pwm_setup
+ *
+ * Description:
+ *   Initialize PWM and register the PWM device.
+ *
```
Comment (saramonteiro) [inline boards/arm/stm32/nucleo-f446re/src/nucleo-f446re.h:338]: Please add the return description

File: boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: stm32_pwm_setup
+ *
+ * Description:
+ *   Initialize PWM and register the PWM device.
+ *
```
Comment (saramonteiro) [inline boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c:49]: return description

File: boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+  struct pwm_lowerhalf_s *pwm;
+  int ret;
+
+  /* Call stm32_pwminitialize() to get an instance of the PWM interface */
+
+#if defined(CONFIG_STM32_TIM1_PWM)
+  pwm = stm32_pwminitialize(1);
+  if (!pwm)
+    {
+      aerr("ERROR: Failed to get the STM32 PWM lower half\n");
```
Comment (saramonteiro) [inline boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c:64]:
```suggestion
      pwmerr("ERROR: Failed to get the STM32 PWM lower half\n");
```

File: boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+      return -ENODEV;
+    }
+
+  ret = pwm_register("/dev/pwm0", pwm);
+  if (ret < 0)
+    {
+      aerr("ERROR: pwm_register failed: %d\n", ret);
```
Comment (saramonteiro) [inline boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c:71]:
```suggestion
      pwmerr("ERROR: pwm_register failed: %d\n", ret);
```

File: boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+      return ret;
+    }
+#endif
+
+#if defined(CONFIG_STM32_TIM2_PWM)
+  pwm = stm32_pwminitialize(2);
+  if (!pwm)
+    {
+      aerr("ERROR: Failed to get the STM32 PWM lower half\n");
```
Comment (saramonteiro) [inline boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c:80]:
```suggestion
      pwmerr("ERROR: Failed to get the STM32 PWM lower half\n");
```

File: boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+      return -ENODEV;
+    }
+
+  ret = pwm_register("/dev/pwm1", pwm);
+  if (ret < 0)
+    {
+      aerr("ERROR: pwm_register failed: %d\n", ret);
```
Comment (saramonteiro) [inline boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c:87]:
```suggestion
      pwmerr("ERROR: pwm_register failed: %d\n", ret);
```

File: boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+      return ret;
+    }
+#endif
+
+#if defined(CONFIG_STM32_TIM3_PWM)
+  pwm = stm32_pwminitialize(3);
+  if (!pwm)
+    {
+      aerr("ERROR: Failed to get the STM32 PWM lower half\n");
```
Comment (saramonteiro) [inline boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c:96]:
```suggestion
      pwmerr("ERROR: Failed to get the STM32 PWM lower half\n");
```

File: boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+      return -ENODEV;
+    }
+
+  ret = pwm_register("/dev/pwm2", pwm);
+  if (ret < 0)
+    {
+      aerr("ERROR: pwm_register failed: %d\n", ret);
```
Comment (saramonteiro) [inline boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c:103]:
```suggestion
      pwmerr("ERROR: pwm_register failed: %d\n", ret);
```

File: boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+      return ret;
+    }
+#endif
+
+#if defined(CONFIG_STM32_TIM4_PWM)
+  pwm = stm32_pwminitialize(4);
+  if (!pwm)
+    {
+      aerr("ERROR: Failed to get the STM32 PWM lower half\n");
```
Comment (saramonteiro) [inline boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c:112]:
```suggestion
      pwmerr("ERROR: Failed to get the STM32 PWM lower half\n");
```

File: boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+      return -ENODEV;
+    }
+
+  ret = pwm_register("/dev/pwm3", pwm);
+  if (ret < 0)
+    {
+      aerr("ERROR: pwm_register failed: %d\n", ret);
```
Comment (saramonteiro) [inline boards/arm/stm32/nucleo-f446re/src/stm32_pwm.c:119]:
```suggestion
      pwmerr("ERROR: pwm_register failed: %d\n", ret);
```

**Comment by michallenc** [comment]
@saramonteiro Thanks, done.

**Comment by saramonteiro** [comment]
> @saramonteiro Thanks, done.

Good, could you also remove the SPI from the defconf and add the return description of the functions that I pointed, pls?

**Comment by michallenc** [comment]
> Good, could you also remove the SPI from the defconf and add the return description of the functions that I pointed, pls?

Yeah, sorry for that SPI, I missed that in the previous review. Those return value descriptions vere added though.

**Comment by saramonteiro** [comment]
LGTM

**Pull request merged**